### ISSUE TITLE
Extremely poor performance

### ISSUE THREAD

**Issue by dyc3** [taylorh140/typst-pintora#2]
I've started using this for diagrams in my docs. Previously, I was using mermaid and had a script to render all the figures to svgs ahead of time. Since I've started using this package, the UX is way better, but my build times have shot through the roof. My document only has a few diagrams, and it takes at least 100x longer to build.

```bash
$ time just build
typst compile --font-path ./fonts main.typ main.pdf
just build  96.12s user 0.18s system 100% cpu 1:36.24 total
```

**Comment by taylorh140** [comment]
Yes, this is a known issue. It might be fixed some day but is largely due to the fact that the wasm binary's are being interperted using wasmi. I nearly didn't publish this package due to the performance. Luckly, incremental builds should suffer less even though it still isn't fun. Someday this could be fixed upstream and the performance may increase about 100x after a transfer to wasmer. However, this is a tricky implementation,  so don't expect it too soon. 

This:

https://discord.com/channels/1054443721975922748/1054443722592497796/1197571746061090897

kind of explains it. but there are many similar post about the wasmi performance.

I'll leave this issue open until some magical day when the transfer is done.

**Comment by dyc3** [comment]
I'm not in that discord, would you mind posting the text?

**Comment by taylorh140** [comment]
```
And wasmi is probably gonna be replaced with wasmer/something similar in the CLI and native browser wasm execution in the webapp 
Both are gonna be (a lot) faster than  wasmi (which is currently used in both webapp and CLI)
The main blocker for that being the webapp which needs to use the wasm interpreter of the browser which, according to laurmaedje, is a bit of work (and closed source)
```

**Comment by SantaHey** [comment]
Is there any news about that? @taylorh140

**Comment by taylorh140** [comment]
diagrams should parse in parallel in typst 0.13, which should make this slightly more bearable.

**Comment by Marcono1234** [comment]
Out of curiosity, is part of the problem that this package here bundles the JavaScript code and then uses [jogs](https://typst.app/universe/package/jogs/) to execute that JavaScript code within Typst?

Could performance maybe be increased by compiling the JavaScript code to WASM and then directly loading it as Typst plugin?

Though it seems there are not that many tools which support JavaScript to WASM compilation. There is https://github.com/bytecodealliance/javy, but if I understand it correctly it still bundles the JavaScript code and a JavaScript runtime so it is unclear if that would actually improve performance, and also its output for Pintora seems to be multiple MBs large. Effectively it might be pretty similar to what the jogs package is already doing.

Another option might be to compile the JavaScript code to some other language first, for example to C, and then compile that to WASM, for example with Emscripten. But so far I haven't found an easy way to do this either. It seems someone else even asked this specifically in the context of typst-pintora for QuickJS NG, see https://github.com/quickjs-ng/quickjs/discussions/308, but it seems the discussion then drifted to a different topic.

**Comment by taylorh140** [comment]
@Marcono1234 Yeah, I think this could be a good investment in time, at least for this particular package. I would like to take this route if I can get it to work. Really anything that helps at this point. It's mostly just finding the time to see if this could work or help.

**Comment by coolCucumber-cat** [comment]
I'm having a similar issue. The performance is atrociously bad even for a really simple diagram. It takes 10 seconds for a 20 line sequence diagram. I'm on Typst 0.13.1 on Linux. The live editor [https://pintorajs.vercel.app/demo/live-editor] is really fast in comparison.

**Comment by taylorh140** [comment]
I would like to fix this. but I haven't gotten to a place where I can get "javy" or something similar to compile the whole thing. It's a skill issue.  I might have some time over the holidays to play around with it to see if that will work at all, or increase the performance.
